File: src/main/java/mapping/DtoToModelMapper.java
package mapping;

import hu.bme.aut.trafficsigns.api.response.DetectionResult;
import hu.bme.aut.trafficsigns.api.response.model.Detection;
import hu.bme.aut.trafficsigns.model.DetectedSign;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;

import java.util.List;
import java.util.stream.Collectors;

@Mapper
public interface DtoToModelMapper {
    DtoToModelMapper INSTANCE = Mappers.getMapper(DtoToModelMapper.class);

    @Mapping(source = "sign.signClass", target="classification.serial")
    Detection modelToDto(DetectedSign sign);

    List<Detection> modelToDto(List<DetectedSign> sign);

    @Mapping(source = "detection.classification.serial", target="signClass")
    DetectedSign dtoToModel(Detection detection, Double lat, Double lon);

    default List<DetectedSign> dtoToModel(DetectionResult result, Double lat, Double lon) {
        return result.getObjects().stream().map(
                detection -> dtoToModel(detection, lat, lon)
                ).collect(Collectors.toList());
    }

}

File: src/main/kotlin/hu/bme/aut/trafficsigns/mapping/DtoToModelMapper.kt
package hu.bme.aut.trafficsigns.mapping

import hu.bme.aut.trafficsigns.api.response.DetectionResult
import hu.bme.aut.trafficsigns.api.response.model.Detection
import hu.bme.aut.trafficsigns.api.response.model.LocatedDetection
import hu.bme.aut.trafficsigns.model.DetectedSign
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.factory.Mappers

@Mapper
interface DtoToModelMapper {
    @Mapping(source = "sign.signClass", target = "classification.serial")
    fun modelToDetection(sign: DetectedSign): Detection
    fun modelToDetection(sign: List<DetectedSign>): List<Detection>

    @Mapping(source = "sign.signClass", target = "classification.serial")
    fun modelToLocatedDetection(sign: DetectedSign): LocatedDetection
    fun modelToLocatedDetection(sign: List<DetectedSign>): List<LocatedDetection>

    @Mapping(source = "detection.classification.serial", target = "signClass")
    fun detectionToModel(detection: Detection, lat: Double, lon: Double): DetectedSign

    companion object {
        val INSTANCE = Mappers.getMapper(DtoToModelMapper::class.java)

        fun detectionToModel(result: DetectionResult, lat: Double, lon: Double): List<DetectedSign> {
            return result.objects.map { detection: Detection -> INSTANCE.detectionToModel(detection, lat, lon) }
        }
    }
}